pkg/distributor/distributor.go

```diff
@@ -1347,41 +1340,66 @@ func (d *Distributor) push(ctx context.Context, pushReq *Request) error {
 		tenantRing = d.ingestersRing.ShuffleShard(userID, d.limits.IngestionTenantShardSize(userID))
 	}
 
-	// we must not re-use buffers now until all DoBatch goroutines have finished,
-	// so set this flag false and pass cleanup() to DoBatch.
+	// we must not re-use buffers now until all writes to backends (e.g. ingesters) have completed, which can happen
+	// even after this function returns. For this reason, it's unsafe to cleanup in the defer and we'll do the cleanup
+	// once all backend requests have completed (see cleanup function passed to sendWriteRequestToBackends()).
 	cleanupInDefer = false
 
-	err = ring.DoBatchWithOptions(ctx, ring.WriteNoExtend, tenantRing, keys,
+	return d.sendWriteRequestToBackends(ctx, userID, req, keys, initialMetadataIndex, tenantRing, func() {
+		// This cleanup function is called once all requests to all backends have been completed.
+		// At this point is safe to finally release the original pushReq resources back to the pool.
+		pushReq.CleanUp()
+	})
+}
+
+func (d *Distributor) sendWriteRequestToBackends(ctx context.Context, tenantID string, req *mimirpb.WriteRequest, keys []uint32, initialMetadataIndex int, tenantRing ring.DoBatchRing, cleanup func()) error {
+	// Use an independent context to make sure all backend instances (e.g. ingesters) get samples even if we return early.
+	// It will still take a while to lookup the ring and calculate which instance gets which series,
+	// so we'll start the remote timeout once the first callback is called.
+	remoteRequestContextAndCancel := sync.OnceValues(func() (context.Context, context.CancelFunc) {
+		return context.WithTimeout(context.WithoutCancel(ctx), d.cfg.RemoteTimeout)
+	})
+
+	remoteRequestContext := func() context.Context {
+		ctx, _ := remoteRequestContextAndCancel()
+		return ctx
+	}
```

Comment:
@pstibrany I agree with Oleg, but what are your thoughts? You added this function in https://github.com/grafana/mimir/pull/7871.

Reply:
It was to simplify the code and remove comment saying that we don't cancel here, only use the context. It makes more sense in "full" PR #7871, I'd keep it since we know more changes are coming.

Reply:
> It was to simplify the code and remove comment saying that we don't cancel here, only use the context.

Specifically this comment: 

```
			// Do not cancel the remoteRequestContext in this callback:
 			// there are more callbacks using it at the same time.
```

which is still in this PR :)

Reply:
👍